### ISSUE TITLE
Some implementations details

### ISSUE THREAD

**Issue by UdonDa** [ldq9526/ARShadowGAN#1]
Hi, @ldq9526 

I'm implementing your method in PyTorch.

I have some questions about ARShadowGAN because I can't re-implement....

1. In your paper. you write `while labeled clue images are normalized to`. Is this mean that the masks of the real-world occluder and the real-world shadow (Fig2 (d, e)) are normalized [0, 1]?
2. Then, rShadow and Occuluder Decoder generate masks, which have [0, 1] range?
3. The input image has [-1, 1] range, so do the  vShadow Decoder and refinement Network produce a mask has a range [-1, 1]?
4. How do you add the output of Refinement Network and an input image? Just spatial suck like `out = generated + input`?

Thank you!

**Comment by ldq9526** [comment]
> Hi, @ldq9526
> 
> I'm implementing your method in PyTorch.
> 
> I have some questions about ARShadowGAN because I can't re-implement....
> 
> 1. In your paper. you write `while labeled clue images are normalized to`. Is this mean that the masks of the real-world occluder and the real-world shadow (Fig2 (d, e)) are normalized [0, 1]?
> 2. Then, rShadow and Occuluder Decoder generate masks, which have [0, 1] range?
> 3. The input image has [-1, 1] range, so do the  vShadow Decoder and refinement Network produce a mask has a range [-1, 1]?
> 4. How do you add the output of Refinement Network and an input image? Just spatial suck like `out = generated + input`?
> 
> Thank you!

Thank you for your interest in our work!
The masks of the real-world occluder and the real-world shadow are normalized [0, 1], and correspondingly, the outputs of rShadow and Occuluder Decoder also have the same range [0, 1]. These masks and outputs are taken in the form of an single-channel attention map.
The direct output of Refinement Network is a 256x256x64 feature map, followed by a conv layer which outputs 256x256x3. Then the output and input image perform element-wise addition.

**Comment by UdonDa** [comment]
Thank you for your replying!!

Let me ask you detailed questions about the parts I don't understand.

>The direct output of Refinement Network is a 256x256x64 feature map, followed by a conv layer which outputs 256x256x3

Does it mean that the structure of refinement net such like
input (256x256x64) -> block1 256x256x64 -> ... -> block4 256x256x64 -> output + tanh 256x256x3  ?

Then, according to your weight node, after final element-wise addition, the output image is applied tanh, why?
I think that the original input image will have tanh applied twice.


![スクリーンショット 2021-08-15 15 22 05](https://user-images.githubusercontent.com/25411643/129469234-d92e7176-268d-4e2a-ad9d-1641ea8406d7.png)


![スクリーンショット 2021-08-15 15 31 34](https://user-images.githubusercontent.com/25411643/129469475-9fdb470a-e0ae-4d24-b89c-fba4db393eab.png)

**Comment by ldq9526** [comment]
> Thank you for your replying!!
> 
> Let me ask you detailed questions about the parts I don't understand.
> 
> > The direct output of Refinement Network is a 256x256x64 feature map, followed by a conv layer which outputs 256x256x3
> 
> Does it mean that the structure of refinement net such like
> input (256x256x64) -> block1 256x256x64 -> ... -> block4 256x256x64 -> output + tanh 256x256x3 ?
> 
> Then, according to your weight node, after final element-wise addition, the output image is applied tanh, why?
> I think that the original input image will have tanh applied twice.
> 
> ![スクリーンショット 2021-08-15 15 22 05](https://user-images.githubusercontent.com/25411643/129469234-d92e7176-268d-4e2a-ad9d-1641ea8406d7.png)
> 
> ![スクリーンショット 2021-08-15 15 31 34](https://user-images.githubusercontent.com/25411643/129469475-9fdb470a-e0ae-4d24-b89c-fba4db393eab.png)

The structure of the whole virtual shadow generator can be found in our supplementary material:
http://graphvision.whu.edu.cn/papers/liudaquan2020_supp.pdf
<img width="236" alt="structure" src="https://user-images.githubusercontent.com/20538566/129475770-d2cfe080-7777-4aee-bbfd-68969a0863c4.png">
Each layer output and conv kernel size are listed for convenience.

The tanh function is used for making the final image (output_conv + input_image) have the range [-1, 1], which is necessary because the range of output_conv is not sure. The original input image range is mapped from [0, 255] to [-1, 1] not by tanh but linear normalization.
Similarly, the attention block use sigmoid function to make the output attention maps have the range [0, 1].

**Comment by UdonDa** [comment]
Thank you for your reply!

I know the range of variables and network structures.

I don't know why I can not reproduce your results with PyTorch... (I can reproduce your tensorflow weight. :D)

Actually, my results look very poor (See top of two row).

Please let me ask three questions, sorry...

1) Both generators are trained with an Adam optimizer having beta1=0.9, beta2=0.99, and lr=0.00001 (1e-5), right?

2) How did you use the output mask of US(5)? Accoriding to your graph, it seems to not be used.

3) Did you get poor results similar to ours?

![image](https://user-images.githubusercontent.com/25411643/129498089-bc2289a5-4a9c-4a59-a764-6ddc2cd38024.png)

**Comment by ldq9526** [comment]
> Thank you for your reply!
> 
> I know the range of variables and network structures.
> 
> I don't know why I can not reproduce your results with PyTorch... (I can reproduce your tensorflow weight. :D)
> 
> Actually, my results look very poor (See top of two row).
> 
> Please let me ask three questions, sorry...
> 
> 1. Both generators are trained with an Adam optimizer having beta1=0.9, beta2=0.99, and lr=0.00001 (1e-5), right?
> 2. How did you use the output mask of US(5)? Accoriding to your graph, it seems to not be used.
> 3. Did you get poor results similar to ours?
> 
> ![image](https://user-images.githubusercontent.com/25411643/129498089-bc2289a5-4a9c-4a59-a764-6ddc2cd38024.png)

1. beta1=0.9, beta2=0.99, and lr=0.00001, this is right. The beta1 and beta2 parameters are the default value of Adam optimizer in Tensorflow 1.12.0.
2. The outputs of Attention Block (two 256x256x1 attention maps) are concatenated with input image by channel. Then the concatenation result is feed to vShadow Generator.
When the shadow generator is training, the attention block should be fixed.
If possible, try to directly use my Attention Block parameters to train vShadow Generator? (I have no idea whether tf model parameters can be export to torch or not, since the version is old)

**Comment by taif1984** [comment]
hi , please which programming language used to implement code please help me
